pkg/k8s/watchers/cilium_endpoint_cleanup.go

```diff
@@ -0,0 +1,127 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package watchers
+
+import (
+	"context"
+	"fmt"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/cilium/cilium/pkg/controller"
+	"github.com/cilium/cilium/pkg/k8s"
+	ciliumK8sTypes "github.com/cilium/cilium/pkg/k8s/types"
+	k8sUtils "github.com/cilium/cilium/pkg/k8s/utils"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+)
+
+func (k *K8sWatcher) markForDeletion(cep *ciliumK8sTypes.CiliumEndpoint) {
+	k.ciliumEndpointCleanupMU.Lock()
+	defer k.ciliumEndpointCleanupMU.Unlock()
+	if !k.ciliumEndpointsCleanupEnabled {
+		return
+	}
+	k.markedForGC[cep.Namespace+"/"+cep.Name] = cep
+}
+
+func (k *K8sWatcher) unmarkForDeletionLocked(name string) {
+	delete(k.markedForGC, name)
+}
+
+func (k *K8sWatcher) unmarkForDeletion(name string) {
+	k.ciliumEndpointCleanupMU.Lock()
+	defer k.ciliumEndpointCleanupMU.Unlock()
+	k.unmarkForDeletionLocked(name)
+}
+
+func (k *K8sWatcher) sweep(ctx context.Context) error {
+	k.ciliumEndpointCleanupMU.RLock()
+	// We only will sweep the endpoints that are marked in this instant.
+	markedNow := map[string]*ciliumK8sTypes.CiliumEndpoint{}
+	for key, cep := range k.markedForGC {
+		markedNow[key] = cep
+	}
+	k.ciliumEndpointCleanupMU.RUnlock()
+	for key, cep := range markedNow {
+		// Ensure CEP still exists in cache.
+		if option.Config.EnableCiliumEndpointSlice {
+			k.ciliumEndpointSliceStoreMU.RLock()
+			_, err := k.ciliumEndpointSliceStore.ByIndex("localNode", key)
+			k.ciliumEndpointSliceStoreMU.RUnlock()
+			if err != nil {
+				log.WithField("ciliumEndpoint", key).Debug("during sweep, CEP was not found in local CiliumEndpointSlice store, unmarking")
+				k.unmarkForDeletion(key)
+				continue
+			}
+		} else {
+			k.ciliumEndpointStoreMU.RLock()
+			_, exists, err := k.ciliumEndpointStore.GetByKey(key)
+			k.ciliumEndpointStoreMU.RUnlock()
+			if err != nil {
+				return fmt.Errorf("could not get local ciliumendpoint by key from indexer: %w", err)
+			}
+			if !exists {
+				log.WithField("ciliumEndpoint", key).Debug("during sweep, CEP was not found in local CiliumEndpoint store, unmarking")
+				k.unmarkForDeletion(key)
+				continue
+			}
+		}
+
+		if err := k.endpointManager.OnMissingEndpoint(cep.Namespace, cep.Name, func() error {
+			log.WithField("ciliumEndpoint", key).Debug("deleting stale CiliumEndpoint")
+			k.ciliumEndpointCleanupMU.Lock()
+			defer k.ciliumEndpointCleanupMU.Unlock()
+			if err := k.client.CiliumEndpoints(cep.Namespace).Delete(ctx, cep.Name, meta_v1.DeleteOptions{}); err != nil {
```

Comment:
> This is a little dangerous. In statefulsets, pods will have the same name even if they are deleted and recreated. Thus, there could a race condition where:

Good point, I think this could be fixed by just adding a unique id. 

> I don't think we need the logic in this file because [Cilium Operator is GC CEPs](https://github.com/cilium/cilium/blob/70a531fe2597cf97a61148c72b463a93c871da84/operator/cmd/k8s_cep_gc.go#L25-L35) and [Cilium deletes the CiliumEndpoint when the local endpoint is gone](https://github.com/cilium/cilium/blob/bd3ad0f1e91b617e8b31ca704fc716e661739809/pkg/k8s/watchers/endpointsynchronizer.go#L382-L390). We only need to check and cleanup stale CEPs when the agent starts.

This covers any cases I can think of. However, for the endpoint sync delete to happen the Agent would've had to have been managing the endpoint at some point. If a CEP update/create causes it to get into that state Cilium won't know to clean it up. However, right now the only way for that to happen is with a manual update/create which may not be something we need to correct.

Removed the periodic gc cleanup in latest commit, will squash down later.